File: Algorithms/Searching_Traversal/binarySearch.js
const arr = [1, 3, 5, 7, 8, 9];
const x = 8;

function iterative(arr, value) {
  let start = 0;
  let end = arr.length - 1;

  while (start <= end) {
    let mid = Math.floor((start + end) / 2);

    if (arr[mid] === value) {
      return true;
    } else if (arr[mid] < value) {
      start = mid + 1;
    } else {
      start = mid - 1;
    }
  }
  return false;
}

function recursive(arr, value, start, end) {
  if (start > end) {
    return false;
  }

  let mid = Math.floor((start + end) / 2);

  if (arr[mid] === value) {
    return true;
  }

  if (arr[mid] > value) {
    return recursive(arr, value, start, mid - 1);
  } else {
    return recursive(arr, value, mid + 1, end);
  }
}

console.log(iterative(arr, x));
console.log(iterative(arr, x, 0, arr.length - 1));

File: Algorithms/Searching_Traversal/binarySearch.java
package Algorithms.Searching_Traversal;

// Binary Search in Java

class binarySearch {

    int binarySearchIterative(int array[], int x, int low, int high) {

        // Repeat until the pointers low and high meet each other
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (array[mid] == x)
                return mid;

            if (array[mid] < x)
                low = mid + 1;

            else
                high = mid - 1;
        }

        return -1;
    }

    int binarySearchRecursive(int array[], int x, int low, int high) {

        if (high >= low) {
            int mid = low + (high - low) / 2;

            // If found at mid, then return it
            if (array[mid] == x)
                return mid;

            // Search the left half
            if (array[mid] > x)
                return binarySearchRecursive(array, x, low, mid - 1);

            // Search the right half
            return binarySearchRecursive(array, x, mid + 1, high);
        }

        return -1;
    }

    public static void main(String args[]) {

        binarySearch ob = new binarySearch();
        int array[] = { 3, 4, 5, 6, 7, 8, 9 };
        int n = array.length;
        int x = 8;

        int resultI = ob.binarySearchIterative(array, x, 0, n - 1);
        if (resultI == -1)
            System.out.println("Not found");
        else
            System.out.println("Element found at index " + resultI + " using Iterative Method");

        int resultR = ob.binarySearchRecursive(array, x, 0, n - 1);
        if (resultR == -1)
            System.out.println("Not found");
        else
            System.out.println("Element found at index " + resultR + " using Recursive Method");
    }
}